AVX-58676 Adding support for Spoke Transit Attachment

sairane10 opened the pull request:
Adding the following changes for Megaport peerings and attachment. Also adding support for megaport cloud type in TF

Transit to transit peerings

- gateway1_logical_ifnames
- gateway2_logical_ifnames

Edge spoke to transit attachements

- spoke_gateway_logical_ifnames
- transit_gateway_logical_ifnames

Spoke to transit attachments

- transit_gateway_logical_ifnames

```
resource "aviatrix_transit_gateway_peering" "test_transit_gateway_peering" {
  transit_gateway_name1                       = "edge-transit-megaport-2"
  transit_gateway_name2                       = "edge-transit-megaport-3"
  enable_peering_over_private_network = true
  jumbo_frame = false
  insane_mode = true
  gateway1_logical_ifnames = ["wan1"]
  gateway2_logical_ifnames = ["wan1"]
}

resource "aviatrix_transit_gateway_peering" "test_transit_gateway_peering_2" {
  transit_gateway_name1                       = "edge-transit-megaport-2"
  transit_gateway_name2                       = "aws-transit"
  enable_peering_over_private_network = true
  jumbo_frame = false
  insane_mode = true
  gateway1_logical_ifnames = ["wan1"]
}

resource "aviatrix_edge_spoke_transit_attachment" "test_edge_spoke_transit_attachment" {
  spoke_gw_name   = "saileeMegaport32"
  transit_gw_name = "edge-transit-megaport-2"
  enable_over_private_network = true
  enable_jumbo_frame = false
  enable_insane_mode = true
  spoke_gateway_logical_ifnames = ["wan1"]
  transit_gateway_logical_ifnames = ["wan1"]
}

resource "aviatrix_spoke_transit_attachment" "test_spoke_transit_attachment" {
  spoke_gw_name   = "saileeMegaport3"
  transit_gw_name = "edge-transit-megaport-2"
  enable_over_private_network = true
  enable_jumbo_frame = false
  enable_insane_mode = true
  transit_gateway_logical_ifnames = ["wan1"]
}
```

Commit:
Updating the config for peering connections

Commit:
Adding UT

Commit:
Adding edge spoke to transit attachment

Commit:
Add UT for peerings and attachments

Commit:
Adding the megaport account support

Commit:
Adding precommit fix

stewartbutler commented on aviatrix/resource_aviatrix_edge_spoke_transit_attachment.go:
```diff
@@ -142,6 +153,20 @@ func resourceAviatrixEdgeSpokeTransitAttachmentCreate(ctx context.Context, d *sc
+	if err != nil {
+		return diag.Errorf("failed to get spoke gateway details: %v", err)
+	}
+	if edgeSpoke.CloudType == goaviatrix.EDGEMEGAPORT {
```
Don't you need to use the goaviatrix.IsCloudType helper here? I thought EDGEMEGAPORT is a bitmask?

sairane10 replied:
edgeSpoke.CloudType is an int and goaviatrix.EDGEMEGAPORT is an untyped constant with this value `EDGEMEGAPORT = 1048576`
The above comparison does not throw any error

stewartbutler commented on aviatrix/resource_aviatrix_edge_spoke_transit_attachment.go:
```diff
@@ -158,9 +183,8 @@ func resourceAviatrixEdgeSpokeTransitAttachmentCreate(ctx context.Context, d *sc
-	var err error
 	for i := 0; ; i++ {
-		err = client.CreateSpokeTransitAttachment(attachment)
+		err := client.CreateSpokeTransitAttachment(context.Background(), attachment)
```
Does this need a context with timeout, or some other context?

sairane10 replied:
passed the existing context from the resource create function

stewartbutler commented on aviatrix/resource_aviatrix_edge_spoke_transit_attachment.go:
```diff
@@ -251,50 +275,61 @@ func resourceAviatrixEdgeSpokeTransitAttachmentRead(ctx context.Context, d *sche
 	if attachment.EnableInsaneMode {
 		d.Set("insane_mode_tunnel_number", attachment.InsaneModeTunnelNumber)
 	}
-	if attachment.DstWanInterfaces != "" {
```
This removal is backward-compatible?

sairane10 replied:
dst_wan_interfaces and src_wan_interfaces are not new attributes added for edge gateway support. These are not yet available to users. They will be added for the first time in 8.0 release

stewartbutler commented on aviatrix/resource_aviatrix_edge_spoke_transit_attachment.go:
```diff
@@ -251,50 +275,61 @@ func resourceAviatrixEdgeSpokeTransitAttachmentRead(ctx context.Context, d *sche
 			return diag.Errorf("could not set spoke_prepend_as_path: %v", err)
 		}
-	} else {
-		d.Set("spoke_prepend_as_path", nil)
```
Is this unnecessary because the default is nil?

sairane10 replied:
Yes, there is no need to explicitly set this value to nil

stewartbutler commented on aviatrix/resource_aviatrix_edge_spoke_transit_attachment.go:
```diff
@@ -251,50 +275,61 @@ func resourceAviatrixEdgeSpokeTransitAttachmentRead(ctx context.Context, d *sche
-		if if0.Type == "WAN" {
-			defaultWanInterfaces = append(defaultWanInterfaces, if0.IfName)
+
+	if edgeSpoke.CloudType == goaviatrix.EDGEMEGAPORT {
```
Needs a check function, right?

sairane10 replied:
Updated it to use `goaviatrix.IsCloudType(edgeSpoke.CloudType, goaviatrix.EDGEMEGAPORT)`

stewartbutler commented on aviatrix/resource_aviatrix_edge_spoke_transit_attachment.go:
```diff
@@ -380,3 +415,46 @@ func resourceAviatrixEdgeSpokeTransitAttachmentDelete(ctx context.Context, d *sc
 	return nil
 }
+
+func getDstWanInterfaces(
```
Move to helper file, add tests

sairane10 replied:
Moved the helper function to test file and added UT

stewartbutler commented on aviatrix/resource_aviatrix_edge_spoke_transit_attachment.go:
```diff
@@ -380,3 +415,46 @@ func resourceAviatrixEdgeSpokeTransitAttachmentDelete(ctx context.Context, d *sc
+	return dstWanInterfaceStr, nil
+}
+
+func getEdgeTransitLogicalIfNames(d *schema.ResourceData, client *goaviatrix.Client, attachment *goaviatrix.SpokeTransitAttachment) error {
```
Move to helper file, add tests

sairane10 replied:
Moved the helper function to test file and added UT

stewartbutler commented on aviatrix/resource_aviatrix_edge_spoke_transit_attachment_test.go:
```diff
@@ -80,6 +101,110 @@ resource "aviatrix_edge_spoke_transit_attachment" "test" {
 }
 
+func testAccEdgeSpokeTransitAttachmentConfigEdge(accountName, spokeGwName, spokeSiteID, path, transitGwName, transitSiteID string) string {
+	return fmt.Sprintf(`
```
Move this to testdata and read it in, rather than have the TF code embedded

sairane10 replied:
This change has already been added in this PR - https://github.com/AviatrixSystems/terraform-provider-aviatrix/pull/2162

stewartbutler commented on aviatrix/resource_aviatrix_spoke_transit_attachment.go:
```diff
@@ -117,7 +131,7 @@ func resourceAviatrixSpokeTransitAttachmentCreate(d *schema.ResourceData, meta i
 	for {
 		try++
-		err := client.CreateSpokeTransitAttachment(attachment)
+		err := client.CreateSpokeTransitAttachment(context.Background(), attachment)
```
Does this need a timeout? Seems like the retry logic would pair well with it.

sairane10 replied:
Added the timeout change in this PR - https://github.com/AviatrixSystems/terraform-provider-aviatrix/pull/2162

stewartbutler commented on aviatrix/resource_aviatrix_spoke_transit_attachment_test.go:
```diff
@@ -104,6 +114,94 @@ resource "aviatrix_spoke_transit_attachment" "test" {
 }
 
+func testAccSpokeTransitAttachmentConfigEdge(accountName, spokeGwName, path, transitGwName, transitSiteID string) string {
+	return fmt.Sprintf(`
```
Same as above -- move to testdata

sairane10 replied:
Moved the config to test-data in this PR - https://github.com/AviatrixSystems/terraform-provider-aviatrix/pull/2162

stewartbutler commented on aviatrix/resource_aviatrix_transit_gateway_peering_test.go:
```diff
@@ -132,6 +153,99 @@ func tesAccCheckTransitGatewayPeeringExists(n string) resource.TestCheckFunc {
 }
 
+func testAccTransitGatewayPeeringConfigEdge(accountName, transit1GwName, transit1SiteID, path, transit2GwName, transit2SiteID string) string {
+	return fmt.Sprintf(`
```
Same as above, move to testdata

sairane10 replied:
Moved the config to test-data in this PR - https://github.com/AviatrixSystems/terraform-provider-aviatrix/pull/2162

stewartbutler commented on goaviatrix/spoke_transit_attachment.go:
```diff
@@ -11,22 +11,24 @@ import (
+	NoMaxPerformance             bool   `form:"no_max_performance,omitempty" json:"no_max_performance,omitempty"`
+	EdgeWanInterfaces            string `form:"edge_wan_interfaces,omitempty" json:"edge_wan_interfaces,omitempty"`
+	EdgeWanInterfacesResp        []string
+	DstWanInterfaces             string   `form:"dst_wan_interfaces,omitempty" json:"dst_wan_interfaces,omitempty"`
```
This was removed above, wasn't it? Do we have both DstWan and LogicalIfNames now?

sairane10 replied:
The TF schema will only support logical interface names for edge gateways but while making the api call's I still need to send both, the logical ifnames and the orginal dst_wan_interface names. The api for will convert the logical ifnames into dst_wan_interfaces only for Megaport vendor.

Commit:
Adding the UT's and review changes

Commit:
Adding feedback changes for helper functions

Commit:
Updating the transit code

Commit:
Adding the review feedback changes

Commit:
Adding helper file for edge spoke transit attachment

sairane10 commented:
@stewartbutler I have moved the helper function to a new file and added UT. The test-data file had already been added in this PR - https://github.com/AviatrixSystems/terraform-provider-aviatrix/pull/2162 so I have skipped adding that change here. 
Please review this PR and I will then change the base to master before merging.
